sunbeam-python/sunbeam/plugins/validation/plugin.py

```diff
@@ -0,0 +1,473 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import subprocess
+from dataclasses import dataclass
+from datetime import datetime
+from pathlib import Path
+from typing import Any, Dict, List, Optional
+
+import click
+from croniter import croniter
+from packaging.version import Version
+from rich.console import Console
+
+from sunbeam.clusterd.client import Client
+from sunbeam.clusterd.service import ClusterServiceUnavailableException
+from sunbeam.commands.openstack import OPENSTACK_MODEL
+from sunbeam.jobs.juju import JujuHelper, run_sync
+from sunbeam.jobs.plugin import PluginManager
+from sunbeam.plugins.interface.v1.openstack import (
+    OpenStackControlPlanePlugin,
+    TerraformPlanLocation,
+)
+
+LOG = logging.getLogger(__name__)
+console = Console()
+
+PLUGIN_VERSION = "0.0.1"
+MINIMAL_PERIOD = 15 * 60  # 15 minutes in seconds
+TEMPEST_APP_NAME = "tempest"
+TEMPEST_CONTAINER_NAME = "tempest"
+TEMPEST_CHANNEL = "2023.2/edge"
+TEMPEST_VALIDATION_RESULT = "/var/lib/tempest/workspace/tempest-validation.log"
+VALIDATION_PLUGIN_DEPLOY_TIMEOUT = (
+    60 * 60
+)  # 60 minutes in seconds, tempest can take some time to initialized
+
+
+def validated_schedule(schedule: str) -> str:
+    """Validate the schedule config option.
+
+    Return the valid schedule if valid,
+    otherwise Raise a click BadParameter exception.
+    """
+    # Empty schedule is fine; it means it's disabled in this context.
+    if not schedule:
+        return ""
+
+    # croniter supports second repeats, but vixie cron does not.
+    if len(schedule.split()) == 6:
+        raise click.ClickException(
+            "This cron does not support seconds in schedule (6 fields)."
+            " Exactly 5 columns must be specified for iterator expression."
+        )
+
+    # constant base time for consistency
+    base = datetime(2004, 3, 5)
+
+    try:
+        cron = croniter(schedule, base, max_years_between_matches=1)
+    except ValueError as e:
+        msg = str(e)
+        # croniter supports second repeats, but vixie cron does not,
+        # so update the error message here to suit.
+        if "Exactly 5 or 6 columns" in msg:
+            msg = "Exactly 5 columns must be specified for iterator expression."
+        raise click.ClickException(msg)
+
+    # This is a rather naive method for enforcing this,
+    # and it may be possible to craft an expression
+    # that results in some consecutive runs within 15 minutes,
+    # however this is fine, as there is process locking for tempest,
+    # and this is more of a sanity check than a security requirement.
+    t1 = cron.get_next()
+    t2 = cron.get_next()
+    if t2 - t1 <= MINIMAL_PERIOD:
```

Comment:
Exact 15min is acceptable in the charm ([ref](https://opendev.org/openstack/sunbeam-charms/src/branch/main/charms/tempest-k8s/src/utils/validators.py#L76)), so let's keep them aligned.

Reply:
Nice catch! so we should be careful about this because we are maintaining two copy of the same code

Reply:
Indeed.

Reply:
I'm not able to  resolve my comments, so I'm leaving my +1s here.